modules/alia-component/src/qbits/alia/component.clj

```diff
@@ -0,0 +1,131 @@
+(ns qbits.alia.component
+  "This namespace provides a way to "
+  (:require [com.stuartsierra.component :as component]
+            [clojure.spec.alpha         :as s]
+            [qbits.alia                 :as alia]))
+
+(defn illegal!
+  [& msgs]
+  (throw (IllegalArgumentException. ^String (reduce str "" msgs))))
+
+(defn ->cluster
+  "Validate cluster configuration, then build a `Cluster` instance
+   out of it, thanks to alia."
+  [config]
+  (when-not (s/valid? ::alia/cluster-options config)
+    (illegal! (s/explain-str ::alia/cluster-options config)))
+  (alia/cluster config))
+
+(defrecord CassandraRegistry [cluster-opts query-opts configs clusters]
+  component/Lifecycle
+  (start [this]
+    (let [clusters (reduce-kv #(assoc %1 %2 (->cluster %3)) {} configs)]
+      (assoc this :clusters clusters)))
+  (stop [this]
+    (doseq [[_ cluster] clusters]
+      (.close ^com.datastax.driver.core.Cluster cluster))
+    (assoc this :clusters nil)))
+
+(defn make-cassandra-registry
+  "Builder for `ClusterRegistry` instances."
+  [options]
+  (map->CassandraRegistry options))
+
+(defn by-name
+  "Fetch cluster by name from the registry."
+  [registry k]
+  (or (get (:clusters registry) (keyword k))
+      (illegal! "no such cluster: " k)))
+
+(defn keyspace-session
+  "Retrieve a `Session` instance set to use a specific keyspace.
+   keyspace may be a keyword or string."
+  [registry k keyspace]
+  (alia/connect (by-name registry k) keyspace))
+
+(defn session
+  "Retrieve a `Session` instance"
+  [registry k]
+  (alia/connect (by-name registry k)))
+
+(defn close-session
+  "Shut-down a session"
+  [^com.datastax.driver.core.Session session]
+  (.close session))
+
+(defn assert-valid-opts!
+  "Throw when supplied query-options are invalid.
+   Removes nil-keys first to avoid trying to validate removed
+   parameters."
+  [opts]
+  (let [sanitized (reduce-kv #(cond-> %1 %3 (assoc %2 %3)) {} opts)]
+    (when-not (s/valid? ::alia/execute-opts sanitized)
+      (illegal! (s/explain-str ::alia/execute-opts opts)))
+    sanitized))
+
+(defn ->opts
+  "This should provide a solid way to parameterize interaction with
+   Cassandra. For each named query `nick` we determine what options to
+   give to the query in the following way:
+
+   - We fetch defaults from the configuration provided query options
+   - We merge code-local defaults for this query
+   - We merge query-specific defaults from configuration
+   - We add values"
+  [nick globals locals]
+  (let [defaults      (:defaults globals)
+        query-globals (get globals nick)]
+    (-> (merge defaults locals query-globals)
+        (assert-valid-opts!))))
+
+(defn ->query
```

Comment:
Should we make the prepared statement "cache"/registry part of the component? I can imagine at init it would just prepare them all, put them in a component field as a map and allow the use of them directly via a keyword on alia/execute and it's other forms (we can just extend alia/PStatement to allow (component, keyword) tuple/type to statement in registry resolution). This would make the statement 1 map get away from the execution but that might be negligible. 

Still unsure about that one tbh, your approach has also advantages, it's simpler, and doesn't bake the state in the component (which has pro and cons).

Reply:
This reverses the dependency order and makes initialization a bit more clumsy.
One thing that could be done is to hide the `:alia.component/stop` query and store started sessions in an atom to do that at component shutdown.

Reply:
If we want to separate both we could also just have another component for (prepared) statement, that also takes a session in dependency (obviously) and we can then extend PStatement to be able to "execute" them and let component do the prepare part if necessary at system/start, but maybe that's going too far down the rabbit hole, and make a system grow in odd ways for not much advantage.

Reply:
It would make systems grow indeed. I am not sure I see a direct advantage to this approach.